### PR TITLE
GeneratePackageConfigurationCommand: Create hierarchical dirs optionally

### PR DESCRIPTION
**Pull request by fviernau**
This supports organizing the files in a hierarchical directory structure
using each package identifier component as path element.

Signed-off-by: Frank Viernau <frank.viernau@here.com>

### CODE REVIEW

File: helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt [hunk 1]
```diff
@@ -77,11 +85,28 @@ internal class GeneratePackageConfigurationsCommand : CliktCommand(
...
         packageConfiguration.writeAsYaml(outputFile)
         println("Wrote a package configuration to '${outputFile.absolutePath}'.")
     }
+
+    private fun getOutputFile(filename: String): File {
+        val relativeOutputFilePath = buildString {
+            fun appendPathElement(pathElement: String) = append(pathElement.ifBlank { "_" }).append("/")
+
+            if (createHierarchicalDirs) {
+                appendPathElement(packageId.type)
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt:99]: Shouldn't we try to make reuse of `Identifier.toPath()` here?

Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt]: Good point. I've generalized it first to make use of it.

File: utils/src/main/kotlin/Extensions.kt [hunk 2]
```diff
@@ -232,7 +232,13 @@ inline fun <K, V, W> Map<K, V>.zipWithDefault(other: Map<K, V>, default: V, oper
...
  * Return the string encoded for safe use as a file name or "unknown", if the string is empty.
  */
-fun String.encodeOrUnknown() = fileSystemEncode().takeUnless { it.isEmpty() } ?: "unknown"
+fun String.encodeOrUnknown(): String = encodeOr("unknown")
+
+/**
+ * Return the string encoded for safe use as a file name or [default] encoded for safe use as a file name, if this
+ * string is empty.
+ */
+fun String.encodeOr(default: String): String = ifEmpty { default }.fileSystemEncode()
```
Comment (sschuberth) [inline utils/src/main/kotlin/Extensions.kt:241]: Note that this is a change in behavior compared to before: Previously, the default string (i.e. "unknown") was taken literally and not encoded. At the example of "unknown" it does not matter, and probably also not for your use-case, but I still think it's a bit misleading as the function name is "encodeOr" which sounds like "encode it or take whatever I provide (without encoding it)".

I'm not saying we should change the implementation, as I actually like to be on the safe side and also encode the `default` string. Also the docs are pretty clear on the implemented behavior, and I cannot think of a better function name right now.

So bottom line, maybe it's enough to just document the behavior change and why it was done / should not matter in most cases in the commit message.

Comment (sschuberth) [reply utils/src/main/kotlin/Extensions.kt:241]: Oh, and can we call `default` also `emptyValue` like you do in the next commit?

File: model/src/main/kotlin/Identifier.kt [hunk 3]
```diff
@@ -114,9 +114,13 @@ data class Identifier(
 
     /**
      * Create a file system path based on the properties of the [Identifier]. All properties are encoded using
-     * [encodeOrUnknown].
+     * [encodeOr] with [emptyValue] as parameter.
      */
-    fun toPath(separator: String = "/") = components.joinToString(separator) { it.encodeOrUnknown() }
+    fun toPath(separator: String = "/", emptyValue: String = "unknown"): String {
+        require(emptyValue.isNotEmpty())
```
Comment (sschuberth) [inline model/src/main/kotlin/Identifier.kt:120]: Shouldn't this check already go to `encodeOr()`? Because if you replace an empty string with an empty string, you could as well omit a call to `encodeOr()` altogether.

File: helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt [hunk 4]
```diff
@@ -77,11 +85,21 @@ internal class GeneratePackageConfigurationsCommand : CliktCommand(
...
-        val outputFile = outputDir.resolve(filename)
+        val outputFile = getOutputFile(filename)
 
         packageConfiguration.writeAsYaml(outputFile)
         println("Wrote a package configuration to '${outputFile.absolutePath}'.")
     }
+
+    private fun getOutputFile(filename: String): File {
+        val relativeOutputFilePath = if (createHierarchicalDirs) {
+            "${packageId.toPath(emptyValue = "_")}/$filename"
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt:96]: Now my next obvious question is: *Why* to we use `_` here but `unknown` in ORT core? Wouldn't it make sense to align the two? I actually like `_` better than `unknown`, I have to admit, but I believe there was a reason why we did not go for `_` in ORT core... but I cannot recall it right now.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt (hunk 4 above)
Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt:96]: Yes, I'm fine with keeping `_` here (and `unknown` in core), I just wanted to get discussion going.

Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt:96]: Do you think adding a command line options as mentioned above is a good idea?

Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/GeneratePackageConfigurationsCommand.kt:96]: No, I think that would over-do it.

BTW, I just created https://github.com/oss-review-toolkit/ort/issues/2987.

File: model/src/main/kotlin/Identifier.kt (hunk 3 above)
Comment (sschuberth) [inline model/src/main/kotlin/Identifier.kt:120]: This should be dropped now from here.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/Identifier.kt (hunk 3 above)
Comment (fviernau) [reply model/src/main/kotlin/Identifier.kt]: Done.

**Review by sschuberth — APPROVED** [review]
